cmd/virt-api/Dockerfile

```diff
@@ -16,15 +16,38 @@
 # Copyright 2017 Red Hat, Inc.
 #
 
-FROM fedora:26
+# builder
+FROM pkotas/golang-centos:latest as builder
 
-MAINTAINER "The KubeVirt Project" <kubevirt-dev@googlegroups.com>
+RUN yum -y update &&\
```

Comment:
The build step and build dependencies should not be coupled with our runtime.

If you want to use containers to build the Kubevirt assets (which is something we want), make a separate build container. That container will only be used for building and it will have ever dependency required to build our assets.

When the build container runs, it should just be working on the source tree. The working directory would be mounted into the build container and that container would be working directly on src tree on the host filesystem.

After the build container exits, the runtime containers can be built using the assets the build container generates.

Reply:
Hi David, can you please be more specific what do you mean? I am little confused, since the current implementation is based on multistage docker build, which does exactly what you said. 

To explain how it works. The builder phase install all the dependencies inside the dockerfile and copies the required go dependencies inside. Than it builds the go binary file and leaves it aside.
The production phase is a second docker image, which install the runtime dependencies, if any, and copies only the build artifact from the builder docker image.

Once the job is finished, the builder image is discarder and only the production image remains.
This results in smaller images and clearly separate the build and production phase from each other. Moreover this a official and recomended method of building the docker images. More on the topic can be found in a nice article about [builder pattern for dockerfiles](https://blog.alexellis.io/mutli-stage-docker-builds/) and in the [official documentation](https://docs.docker.com/engine/userguide/eng-image/multistage-build/).

Reply:
@rmohr can You please explain in more detail what are you missing? To exaplin in more detail.

The builder stage is doing exactly what you mentioned. It install all the required yum dependencies inside the build phase. Next, the golang dependencies are copied inside and the $GOPATH is set for the ubevirt project. Finally the go build is executed and result is waiting for the collection. If I missed some yum dependencie, that is a different question. Easy to fix, though.

The good thing about this approach is that it make use of the docker caching. This means, build phase is reused for all docker images build with the two stage build docker file.

Alsp, I have went through the #447. Altought overcomplicated, it is still nice contribution. Compared to two stage build phase, is add aditional complexity to the build scripts. On the other hand, two stage build system removes the complexity from the build scripts and does not introduce change in the current workflow.

Reply:
@petrkotas I see what this does and I appreciate your enthusiasm and there is nothing wrong with it per se, but I want to recommend to create one Dockerfile which is the builder (has go, yum and go-generate dependencies installed).

Then build the project somehow like this:

```bash
docker run --rm -v "$PWD":"go/src/kubevirt.io/kubevirt" -w go/src/kubevirt.io/kubevirt make generate build
```
No need to touch the other Dockerfiles. All resulting artefacts are at their usual places and "make docker" can just build them with the resulting binaries from before. So one can run

```bash
docker run --rm -v "$PWD":"go/src/kubevirt.io/kubevirt" -w go/src/kubevirt.io/kubevirt make generate build
make docker
```
instead of

```bash
make generate build
make docker
```

This way one can still just run "make" directly to achieve everything, but one can also use the docker build, to not have to install all yum dependencies and golang just for building.

Reply:
Ah, that is a fair point. I will rework the case so make can still be used on bare metal. 
Thank you for clarification of the requirement.

Reply:
@rmohr I have removed the two-stage build for this PR. I think I would like to persist is the switch from fedora to centos for the sake of stability.

Would you be ok with it. In the meantime I will work on #447 .

Reply:
To be honest, I would prefer if we keep it. It was centos before, now it is fedora, but except from libvirt-client (fedora gives us an easy way of getting newest versions), it does almost not matter at all, since we just execute pre-compiled go-binaries on the host kernel.

Reply:
I understand, that using Fedora gives us the bleeding edge, which could be fine for development.
My point in using more stable centos is, that I do not want to fight broken repositories of Fedora. Which I most of the times do when building the images. The repositories tend to get fixed fast, but it is very inconvenient to know, that you are just going to wait for the maintainers to fix it.

Therefore I would like to use Centos since it is more stable, more mature and offers longer support. 

As for the bleeding edge, can you share which library we require to be the bleeding edge for our development? Maybe we can get it some other way, then wait for the repositories. We are using containers, which gives us the opportunity to play a little bit more with dependencies :).

Reply:
Well it is just libvirt-client and libvirt-devel. We were using a copr repo fro centos before. I am not necessarily against going back to centos, but I don't want to permanently switch back and forth. In doubt I am for keeping fedora right now, just because it is what we have, until we know for sure what we want to use. @davidvossel, @fabiand ?